File: ats/ax/classification/confidence.go
package classification

import (
	"math"
	"time"
)

// ConfidenceCalculator calculates confidence scores for conflict resolution
type ConfidenceCalculator struct {
	credibilityManager *CredibilityManager
	temporalAnalyzer   *TemporalAnalyzer
	reviewThreshold    float64 // Below this threshold = human review required
}

// NewConfidenceCalculator creates a new confidence calculator
func NewConfidenceCalculator(cm *CredibilityManager, ta *TemporalAnalyzer) *ConfidenceCalculator {
	return &ConfidenceCalculator{
		credibilityManager: cm,
		temporalAnalyzer:   ta,
		reviewThreshold:    0.3, // Default: 30% confidence threshold
	}
}

// SetReviewThreshold sets the confidence threshold below which human review is required
func (cc *ConfidenceCalculator) SetReviewThreshold(threshold float64) {
	cc.reviewThreshold = threshold
}

// CalculateConfidence calculates overall confidence score for a set of claims
func (cc *ConfidenceCalculator) CalculateConfidence(claims []ClaimWithTiming) float64 {
	if len(claims) == 0 {
		return 0.0
	}

	if len(claims) == 1 {
		return cc.calculateSingleClaimConfidence(claims[0])
	}

	// Multi-claim confidence calculation
	baseScore := 0.5

	// Independent source bonus (+0.3 max)
	sourceBonus := cc.calculateSourceDiversityBonus(claims)

	// Actor credibility bonus (+0.2 max)
	credibilityBonus := cc.calculateCredibilityBonus(claims)

	// Temporal pattern bonus (+0.2 max)
	temporalBonus := cc.calculateTemporalBonus(claims)

	// Recency bonus (+0.1 max)
	recencyBonus := cc.calculateRecencyBonus(claims)

	// Consistency bonus (+0.1 max)
	consistencyBonus := cc.calculateConsistencyBonus(claims)

	totalScore := baseScore + sourceBonus + credibilityBonus + temporalBonus + recencyBonus + consistencyBonus

	return math.Min(totalScore, 1.0)
}

// ClaimWithTiming represents a claim with timing information for confidence calculation
type ClaimWithTiming struct {
	Actor     string
	Timestamp time.Time
	Predicate string
	Subject   string
	Context   string
}

// calculateSingleClaimConfidence calculates confidence for a single claim
func (cc *ConfidenceCalculator) calculateSingleClaimConfidence(claim ClaimWithTiming) float64 {
	// Single claims get confidence based on actor credibility and recency
	credibility := cc.credibilityManager.GetActorCredibility(claim.Actor)
	recency := cc.temporalAnalyzer.CalculateRecencyScore(claim.Timestamp)

	// Weight credibility more heavily for single claims
	return (credibility.Authority * 0.7) + (recency * 0.3)
}

// calculateSourceDiversityBonus calculates bonus for having multiple independent sources
func (cc *ConfidenceCalculator) calculateSourceDiversityBonus(claims []ClaimWithTiming) float64 {
	uniqueActors := make(map[string]bool)
	for _, claim := range claims {
		uniqueActors[claim.Actor] = true
	}

	independentCount := len(uniqueActors)
	if independentCount <= 1 {
		return 0.0
	}

	// Bonus increases with more independent sources, max 0.3
	bonus := float64(independentCount-1) * 0.1
	return math.Min(bonus, 0.3)
}

// calculateCredibilityBonus calculates bonus based on highest credibility actor
func (cc *ConfidenceCalculator) calculateCredibilityBonus(claims []ClaimWithTiming) float64 {
	var actors []string
	for _, claim := range claims {
		actors = append(actors, claim.Actor)
	}

	highest := cc.credibilityManager.GetHighestCredibility(actors)
	return highest.Authority * 0.2 // Max 0.2 bonus for highest credibility actor
}

// calculateTemporalBonus calculates bonus based on temporal patterns
func (cc *ConfidenceCalculator) calculateTemporalBonus(claims []ClaimWithTiming) float64 {
	var timings []ClaimTiming
	for _, claim := range claims {
		timings = append(timings, ClaimTiming{
			Actor:     claim.Actor,
			Timestamp: claim.Timestamp,
			Predicate: claim.Predicate,
		})
	}

	temporalConfidence := cc.temporalAnalyzer.CalculateTemporalConfidence(timings)
	return (temporalConfidence - 0.5) * 0.4 // Convert 0.5-1.0 range to 0.0-0.2 bonus
}

// calculateRecencyBonus calculates bonus based on most recent claim
func (cc *ConfidenceCalculator) calculateRecencyBonus(claims []ClaimWithTiming) float64 {
	var mostRecent time.Time
	for _, claim := range claims {
		if claim.Timestamp.After(mostRecent) {
			mostRecent = claim.Timestamp
		}
	}

	recency := cc.temporalAnalyzer.CalculateRecencyScore(mostRecent)
	return recency * 0.1 // Max 0.1 bonus for recency
}

// calculateConsistencyBonus calculates bonus for consistent predicates
func (cc *ConfidenceCalculator) calculateConsistencyBonus(claims []ClaimWithTiming) float64 {
	uniquePredicates := make(map[string]bool)
	for _, claim := range claims {
		uniquePredicates[claim.Predicate] = true
	}

	// If all claims have same predicate, give consistency bonus
	if len(uniquePredicates) == 1 {
		return 0.1
	}

	// If predicates are different but related (e.g., junior/senior developer), smaller bonus
	if cc.areRelatedPredicates(uniquePredicates) {
		return 0.05
	}

	return 0.0
}

// areRelatedPredicates checks if predicates suggest evolution rather than conflict
func (cc *ConfidenceCalculator) areRelatedPredicates(predicates map[string]bool) bool {
	// Simple heuristic: if one predicate contains another, they might be related
	predicateList := make([]string, 0, len(predicates))
	for p := range predicates {
		predicateList = append(predicateList, p)
	}

	for i, p1 := range predicateList {
		for j, p2 := range predicateList {
			if i != j && (contains(p1, p2) || contains(p2, p1)) {
				return true
			}
		}
	}

	return false
}

// contains checks if s1 contains s2 (case-insensitive substring)
func contains(s1, s2 string) bool {
	// Simple substring check - could be enhanced with more sophisticated logic
	return len(s2) > 0 && len(s1) > len(s2) &&
		(s1[:len(s2)] == s2 || s1[len(s1)-len(s2):] == s2)
}

// RequiresHumanReview returns true if confidence is below review threshold
func (cc *ConfidenceCalculator) RequiresHumanReview(confidence float64) bool {
	return confidence < cc.reviewThreshold
}

// GetConfidenceLevel returns a human-readable confidence level
func (cc *ConfidenceCalculator) GetConfidenceLevel(confidence float64) string {
	switch {
	case confidence >= 0.8:
		return "high"
	case confidence >= 0.6:
		return "medium"
	case confidence >= 0.4:
		return "low"
	default:
		return "very_low"
	}
}

// CalculateActorAgreement calculates how much actors agree on the same claim
func (cc *ConfidenceCalculator) CalculateActorAgreement(claims []ClaimWithTiming) float64 {
	if len(claims) <= 1 {
		return 1.0
	}

	// Group claims by predicate
	predicateGroups := make(map[string][]ClaimWithTiming)
	for _, claim := range claims {
		predicateGroups[claim.Predicate] = append(predicateGroups[claim.Predicate], claim)
	}

	// Find the predicate with most agreement
	var maxAgreement int
	for _, group := range predicateGroups {
		if len(group) > maxAgreement {
			maxAgreement = len(group)
		}
	}

	return float64(maxAgreement) / float64(len(claims))
}

File: crates/qntx-core/src/classify/confidence.rs
//! Confidence scoring for conflict classification
//!
//! Multi-factor confidence calculation:
//! - Source diversity bonus (multiple independent actors)
//! - Actor credibility bonus (highest credibility actor)
//! - Temporal pattern bonus (simultaneous vs distributed)
//! - Recency bonus (how recent the most recent claim is)
//! - Consistency bonus (all claims agree on predicate)

use super::credibility::ActorCredibility;
use super::temporal::{ClaimTiming, TemporalAnalyzer};

/// Claim with full context for confidence calculation
#[derive(Debug, Clone)]
pub struct ClaimWithTiming {
    pub actor: String,
    pub timestamp_ms: i64,
    pub predicate: String,
    pub subject: String,
    pub context: String,
}

impl ClaimWithTiming {
    fn to_timing(&self) -> ClaimTiming {
        ClaimTiming {
            actor: self.actor.clone(),
            timestamp_ms: self.timestamp_ms,
            predicate: self.predicate.clone(),
        }
    }
}

/// Confidence calculator with configurable review threshold
pub struct ConfidenceCalculator<'a> {
    temporal: &'a TemporalAnalyzer,
    review_threshold: f64,
}

impl<'a> ConfidenceCalculator<'a> {
    pub fn new(temporal: &'a TemporalAnalyzer) -> Self {
        Self {
            temporal,
            review_threshold: 0.3,
        }
    }

    pub fn with_review_threshold(mut self, threshold: f64) -> Self {
        self.review_threshold = threshold;
        self
    }

    /// Calculate overall confidence score for a set of claims.
    /// `now_ms` is the current time in milliseconds for recency calculation.
    pub fn calculate(&self, claims: &[ClaimWithTiming], now_ms: i64) -> f64 {
        if claims.is_empty() {
            return 0.0;
        }

        if claims.len() == 1 {
            return self.single_claim_confidence(&claims[0], now_ms);
        }

        let base_score = 0.5;
        let source_bonus = self.source_diversity_bonus(claims);
        let credibility_bonus = self.credibility_bonus(claims);
        let temporal_bonus = self.temporal_bonus(claims);
        let recency_bonus = self.recency_bonus(claims, now_ms);
        let consistency_bonus = self.consistency_bonus(claims);

        let total = base_score
            + source_bonus
            + credibility_bonus
            + temporal_bonus
            + recency_bonus
            + consistency_bonus;

        total.min(1.0)
    }

    /// Whether a confidence score requires human review
    pub fn requires_review(&self, confidence: f64) -> bool {
        confidence < self.review_threshold
    }

    /// Single claim confidence based on actor credibility and recency
    fn single_claim_confidence(&self, claim: &ClaimWithTiming, now_ms: i64) -> f64 {
        let credibility = ActorCredibility::from_actor(&claim.actor);
        let recency = self.temporal.recency_score(claim.timestamp_ms, now_ms);
        (credibility.score() * 0.7) + (recency * 0.3)
    }

    /// Bonus for having multiple independent sources (+0.3 max)
    fn source_diversity_bonus(&self, claims: &[ClaimWithTiming]) -> f64 {
        let mut unique_actors = std::collections::HashSet::new();
        for claim in claims {
            unique_actors.insert(&claim.actor);
        }

        let independent_count = unique_actors.len();
        if independent_count <= 1 {
            return 0.0;
        }

        let bonus = (independent_count - 1) as f64 * 0.1;
        bonus.min(0.3)
    }

    /// Bonus based on highest credibility actor (+0.2 max)
    fn credibility_bonus(&self, claims: &[ClaimWithTiming]) -> f64 {
        let highest = claims
            .iter()
            .map(|c| ActorCredibility::from_actor(&c.actor))
            .max()
            .unwrap_or(ActorCredibility::External);

        highest.score() * 0.2
    }

    /// Bonus based on temporal patterns (+0.2 max)
    fn temporal_bonus(&self, claims: &[ClaimWithTiming]) -> f64 {
        let timings: Vec<ClaimTiming> = claims.iter().map(|c| c.to_timing()).collect();
        let temporal_confidence = self.temporal.temporal_confidence(&timings);
        (temporal_confidence - 0.5) * 0.4 // Convert 0.5-1.0 range to 0.0-0.2
    }

    /// Bonus based on most recent claim (+0.1 max)
    fn recency_bonus(&self, claims: &[ClaimWithTiming], now_ms: i64) -> f64 {
        let timings: Vec<ClaimTiming> = claims.iter().map(|c| c.to_timing()).collect();
        if let Some(most_recent) = self.temporal.most_recent(&timings) {
            let recency = self.temporal.recency_score(most_recent, now_ms);
            recency * 0.1
        } else {
            0.0
        }
    }

    /// Bonus for consistent predicates (+0.1 max)
    fn consistency_bonus(&self, claims: &[ClaimWithTiming]) -> f64 {
        let mut unique_predicates = std::collections::HashSet::new();
        for claim in claims {
            unique_predicates.insert(&claim.predicate);
        }

        if unique_predicates.len() == 1 {
            return 0.1;
        }

        // Check if predicates are related (one contains the other as prefix/suffix)
        if self.are_related_predicates(&unique_predicates) {
            return 0.05;
        }

        0.0
    }

    /// Heuristic: if one predicate contains another, they might be related
    fn are_related_predicates(&self, predicates: &std::collections::HashSet<&String>) -> bool {
        let list: Vec<&&String> = predicates.iter().collect();
        for (i, p1) in list.iter().enumerate() {
            for (j, p2) in list.iter().enumerate() {
                if i != j && (p1.starts_with(p2.as_str()) || p1.ends_with(p2.as_str())) {
                    return true;
                }
            }
        }
        false
    }
}

#[cfg(test)]
mod tests {
    use super::super::temporal::TemporalConfig;
    use super::*;

    fn setup() -> (TemporalAnalyzer, i64) {
        let ta = TemporalAnalyzer::new(TemporalConfig::default());
        let now = 1_000_000_000; // some reference time
        (ta, now)
    }

    fn claim(actor: &str, ts: i64, predicate: &str) -> ClaimWithTiming {
        ClaimWithTiming {
            actor: actor.to_string(),
            timestamp_ms: ts,
            predicate: predicate.to_string(),
            subject: "ALICE".to_string(),
            context: "GitHub".to_string(),
        }
    }

    #[test]
    fn empty_claims_zero_confidence() {
        let (ta, now) = setup();
        let cc = ConfidenceCalculator::new(&ta);
        assert_eq!(cc.calculate(&[], now), 0.0);
    }

    #[test]
    fn single_human_claim_high_confidence() {
        let (ta, now) = setup();
        let cc = ConfidenceCalculator::new(&ta);
        let claims = vec![claim("human:alice", now - 1000, "is_author_of")];
        let score = cc.calculate(&claims, now);
        // Human credibility (1.0 * 0.7) + recency (1.0 * 0.3) = 1.0
        assert!(
            score > 0.9,
            "human claim should be high confidence, got {}",
            score
        );
    }

    #[test]
    fn multiple_sources_boost_confidence() {
        let (ta, now) = setup();
        let cc = ConfidenceCalculator::new(&ta);
        let claims = vec![
            claim("human:alice", now - 1000, "is_author_of"),
            claim("human:bob", now - 2000, "is_author_of"),
            claim("system:ci", now - 3000, "is_author_of"),
        ];
        let score = cc.calculate(&claims, now);
        assert!(
            score > 0.7,
            "multiple sources should boost confidence, got {}",
            score
        );
    }

    #[test]
    fn review_threshold() {
        let (ta, _now) = setup();
        let cc = ConfidenceCalculator::new(&ta).with_review_threshold(0.5);
        assert!(cc.requires_review(0.3));
        assert!(!cc.requires_review(0.6));
    }
}
